044-cluster-wide-volume-usage-quota-management.md

```diff
@@ -0,0 +1,253 @@
+without triggering throttling of clients, as broker `1` disk usage remains acceptable.
+
+Addressing the effects of uneven topic distribution sounds like it should come under [KIP-73](https://cwiki.apache.org/confluence/display/KAFKA/KIP-73+Replication+Quotas). Unfortunately replication quotas
+are designed to manage the additional network load of migrating replicas between brokers, which does not address client publication
+leading to out of disk conditions through replication.
+
+Currently, the kafka-quotas-plugin considers the total quantity of storage and how much of that is consumed (
+see [issue#2](https://github.com/strimzi/kafka-quotas-plugin/issues/2)) when considering whether to apply throttling to
+clients. This is problematic with respect to handling disk failure for Just a Bunch Of Disks (JBOD) deployments [(KIP-112)](https://cwiki.apache.org/confluence/display/KAFKA/KIP-112%3A+Handle+disk+failure+for+JBOD) as the
+broker will take partitions offline when the volume they are stored on runs out of space. Which in the case of unbalanced usage
+between volumes can lead to a volume running out of storage without throttling being applied. The broker will go offline
+if all volumes are unavailable.
+
+Throttling down message production on all broker nodes will protect the cluster from
+running out of disk due to replication.
+
+See also:
+* [kafka-quotas-plugin issue #28 - KIP-827 support](https://github.com/strimzi/kafka-quotas-plugin/issues/28)
+* [kafka-quotas-plugin issue #2 - Improve support for JBOD storage](https://github.com/strimzi/kafka-quotas-plugin/issues/2)
+
+## Motivation
+
+Users need better protection against running out of disk on any of their nodes as it can degrade the service
+in unpredictable ways (corrupted segment logs, affect other users of that disk etc) which could impede
+the interventions required to recover.
+
+## Proposal
+
+1. Extend the kafka-quotas-plugin so that we can choose to observe the disk usage 
+of all brokers in the cluster.
+2. Deprecate the existing local disk observations.
+3. Add new limit types, so that it is explicit what is being limited and better support heterogeneous disks
+   1. throttle if consumed bytes exceeds threshold on any volume
+   2. throttle if usable bytes less than threshold on any volume
+   3. throttle if usable ratio less than threshold on any volume
+4. Introduce an extension point in the quotas-plugin to support pluggable sources of quotas
+5. Introduce a fallback throttle factor in case we cannot retrieve the cluster state
+
+So every broker will make its own independent throttling decision based on knowledge of the volumes on all active broker nodes. If it detects that any volume in
+the cluster is becoming too full it will throttle production of messages. The kafka quota API isn't rich enough to do anything 
+smarter about only throttling writes to the brokers running out of space, so we fence the whole cluster.
```

Comment:
We wanted to point out that the throttling decision is not centralised. The brokers use similar data and make a deterministic decision (similar since they will be polling at different times).

Have added that to the explanation.